### ISSUE TITLE
[explicit-module-boundary-types] Callback error inside exported React function component

### ISSUE THREAD

**Issue by ford04** [typescript-eslint/typescript-eslint#1626]
**Repro**

```JSON
{
  rules: {
    "@typescript-eslint/explicit-function-return-type": "off",
    "@typescript-eslint/explicit-module-boundary-types": ["error"],
  }
}
```
Other rules are omitted for brevity. I can hand them in, if needed.
```TS
import React, { FC } from "react"

export const Comp: FC = () => (
  <div
    onClick={e => {
      console.log("click!", e)
    }}
    // Missing return type on function.
    // Argument 'e' should be typed.
  ></div>
)
```

**Expected Result**
No eslint error is triggered, as the exported outer function `Comp` is explicitely typed as `FC`. My interpretation of the rule is, that only directly exported members should be checked.

**Actual Result**
Two errors are raised:  
```
1) Missing return type on function.eslint(@typescript-eslint/explicit-module-boundary-types)
2) Argument 'e' should be typed.eslint(@typescript-eslint/explicit-module-boundary-types)
```

I am also not sure about the second error, as `e` should be contextually typed by `onClick?: MouseEventHandler<T>;` in  `DOMAttributes<T>`. It seems that this rule wants a fully typed function, even if the function is only part of `Comp` and itself not exported.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.20.0` |
| `@typescript-eslint/parser`        | `2.20.0` |
| `TypeScript`                       | `3.7.5` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `13.9.0` |
| `npm`                              | `6.13.7` |

**Comment by bradzacher** [comment]
The rule doesn't have JSX context encoded. It's a simple enough fix. Happy to accept a PR.

**Comment by ford04** [comment]
@bradzacher thanks for clarifying.

Could you give some more context, what you mean by "doesn't have JSX context encoded"?

I had a look at the [rule source](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/explicit-module-boundary-types.ts) and tried to find a a related example rule involving JSX. [comma-spacing](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/tests/rules/comma-spacing.test.ts) seems to be one, as it uses  `parserOptions: { ecmaFeatures: { jsx: true } }` in some cases. But I am not sure how to proceed here, after having looked at its [implementation](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/comma-spacing.ts).

**Comment by bradzacher** [comment]
If you want to work on this, I'd suggest using the debugger to step through your repro code in the rule.

Add your code as a test case to the test file for the rule, and then use vscode to debug how the ast is processed.
In particular the problem is in the `isUnexported` function.
